### PR TITLE
[NEW] GDPR - Right to access and Data Portability

### LINKED ISSUE #9767: [GDPR] Right to Access
## Description
Part of the expanded rights of data subjects outlined by the GDPR is the right for data subjects to obtain from the data controller confirmation as to whether or not personal data concerning them is being processed, where and for what purpose. Further, the controller shall provide a copy of the personal data, free of charge, in an electronic fromat. This change is a dramatic shift to data transparency and empowerment of data subjects.

## Comment
Maybe a web page for registered users to download their content stored in the database. Check the Facebook feature for ideas, they have solved it in a nice way I think.

## Solution (draft)
- Admin option to enable it
- Profile button to start the process
- Method to loop all the user's subscriptions and create a file per room containing the conversation (all messages or user's messages only?) in human readable format
	- Get all the attachments too
- Notify via email the link to download the zip containing the files
	- Store inside mongodb???
- Only allow to execute the method once per day

## Effort
- 32h of work

### LINKED ISSUE #9768: [GDRP] Data Portability
## Description
GDPR introduces data portability - the right for a data subject to receive the personal data concerning them, which they have previously provided in a 'commonly use and machine readable format' and have the right to transmit that data to another controller. 

## Comment
This is a way to export all available data in order to more easily move to other platforms - if possible. I would suggest that Rocket Chat implements this as JSON export of all user data and make it an option in the Right to Access feature, keep it as simple as possible!

## Solution (draft)
- Admin option to enable it
- Profile button to start the process
- Method to loop all the user's subscriptions and create a file per room containing the conversation (all messages or user's messages only?) in machine format
	- Get all the attachments too
- Notify via email the link to download the zip containing the files
	- Store inside mongodb???
- Only allow to execute the method once per day
- 24h of work

### PR DESCRIPTION
**Pull request by Hudell**
@RocketChat/core 

Closes #9767
Closes #9768 

This PR adds an option for users to download all their data stored in the servers, in either JSON or HTML format. This funcionality can be enabled by an admin through a new page on the administration center.

Administration Center

Once enabled, two new buttons will be added to the user's account page:
My Account

When the user clicks on one of those new buttons, the system will register the request on the new exportOperations collection. The server will then automatically export all the messages and attachments from every channel that the user has access to. When this operation is complete, the server will generate a zip file and send an email to the user with a download link.

Each button can only be used once per day. The download button will generate HTML files and the export button will generate JSON files.

### CODE REVIEW

File: packages/rocketchat-user-data-download/server/lib/requests.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+		if (operation && operation.status === 'completed') {
+			if (operation.generatedFile !== null) {
+
+				if (!Meteor.settings.public.sandstorm && !FileUpload.requestCanAccessFiles(req)) {
+					res.writeHead(403);
+					return res.end();
+				}
+
+				const filePath = operation.generatedFile;
```
Comment (geekgonecrazy) [inline packages/rocketchat-user-data-download/server/lib/requests.js]: We are making a lot of assumptions about long term storage being available on the disk.  Could we store the file on the file store provider already setup on the server?

Let's just imagine that this is in a docker container.  If you write the file and then the container is restarted the file is gone.  Mounting a directory is possible... But in some setups this is pretty difficult

Comment (Hudell) [reply packages/rocketchat-user-data-download/server/lib/requests.js]: I've added GridFS, Amazon and Google Storage as options for the location of the generated file.

**Comment by photoninger** [comment]
```mkdir -p``` would avoid such errors

**Comment by Elvinra** [comment]
On docker same problem, the user rocketchat don't have the permissions on tmp folder
- little fix: Mount a volume on /tmp/userData, put chmod 777 on the host system

Real fix: update the dockerfile to create that folder and put the permissions to rocketchat user

**Comment by photoninger** [comment]
When are the exported files and zip files deleted? 
Or do you depend on deleting everything in /tmp on reboot?

**Pull request merged**